### PR TITLE
Orderby support in entity odata endpoint

### PR DESCRIPTION
**Pull request by ktuite**
Closes #

Adds $orderby support to OData endpoints for Entities and Submissions. This will be useful overall but it is also part of Frontend showing existing entities in the bulk entity creation flow. 

#### What has been done to verify that this works as intended?

Tests

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

New feature for users! We've covered a lot of the OData spec at this point!

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

Documentation has been updated in this PR. 

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

**Comment by matthew-white** [comment]
Is there a way to ensure that the sort is stable? Currently, the `ORDER BY` clause ends with `entities.id DESC`, and I think it'd be useful to continue doing that. Though with `$orderby`, maybe the `DESC`/`ASC` after `entities.id` should match the `DESC`/`ASC` that the user specifies for the first `$orderby` property. 🤔

**Comment by matthew-white** [comment]
I'm going to add this PR to the project even though getodk/central#589 was the main impetus behind it. It's a nice improvement on its own that we'll want to mention in release notes.

File: lib/data/odata-filter.js [hunk 1]
```diff
@@ -87,5 +87,27 @@ const odataFilter = (expr, odataToColumnMap) => {
   return op(ast);
 };
 
-module.exports = { odataFilter };
+const odataOrderBy = (expr, odataToColumnMap, stableOrderColumn = null) => {
+  if (expr == null) return sql`true`;
```
Comment (ktuite) [inline lib/data/odata-filter.js:91]: 'true' doesn't make sense in context

File: lib/data/odata-filter.js [hunk 2]
```diff
@@ -87,5 +87,27 @@ const odataFilter = (expr, odataToColumnMap) => {
...
+
+  let initialOrder = null;
+  const clauses = expr.split(',').map((exp) => {
+    const [col, order] = exp.trim().split(/\s+/);
+    if (!odataToColumnMap.has(col))
+      throw Problem.internal.unsupportedODataField({ at: 0, text: col });
```
Comment (ktuite) [inline lib/data/odata-filter.js:97]: make "at" optional

File: lib/data/odata-filter.js [hunk 3]
```diff
@@ -87,5 +87,27 @@ const odataFilter = (expr, odataToColumnMap) => {
   return op(ast);
 };
 
-module.exports = { odataFilter };
+const odataOrderBy = (expr, odataToColumnMap, stableOrderColumn = null) => {
+  if (expr == null) return sql`true`;
+
+  let initialOrder = null;
+  const clauses = expr.split(',').map((exp) => {
+    const [col, order] = exp.trim().split(/\s+/);
```
Comment (ktuite) [inline lib/data/odata-filter.js:95]: have order default to ascending here

File: lib/data/odata-filter.js [hunk 4]
```diff
@@ -87,5 +87,27 @@ const odataFilter = (expr, odataToColumnMap) => {
...
+    if (!odataToColumnMap.has(col))
+      throw Problem.internal.unsupportedODataField({ at: 0, text: col });
+
+    // Save the order of the initial property to use for the stable sort column order
+    if (initialOrder == null)
+      initialOrder = order;
+
+    return sql`${sql.identifier(odataToColumnMap.get(col).split('.'))} ${((order?.toLowerCase() === 'asc') ? sql`ASC` : sql`DESC`)}`;
```
Comment (ktuite) [inline lib/data/odata-filter.js:103]: odata spec (https://learn.microsoft.com/en-us/azure/search/search-query-odata-orderby) mentions " If there are null values in the field, null values appear first if the sort is asc and last if the sort is desc."

Add `NULLS FIRST` to end of each sort clause.

File: test/unit/data/odata-filter.js [hunk 5]
```diff
@@ -73,3 +74,48 @@ describe('OData filter query transformer', () => {
...
+    odataOrderBy('__system/updatedAt   asc').should.eql(sql`ORDER BY ${sql.identifier([ 'submissions', 'updatedAt' ])} ASC`);
+    odataOrderBy('  __system/updatedAt   AsC ').should.eql(sql`ORDER BY ${sql.identifier([ 'submissions', 'updatedAt' ])} ASC`);
+  });
+
+  it('should combine multiple sort operators', () => {
+    odataOrderBy('__system/updatedAt desc, __system/submissionDate ASC').should.eql(sql`ORDER BY ${sql.identifier([ 'submissions', 'updatedAt' ])} DESC,${sql.identifier([ 'submissions', 'createdAt' ])} ASC`);
+  });
+
+  it('should handle no asc/desc provided', () => {
+    odataOrderBy('__system/updatedAt').should.eql(sql`ORDER BY ${sql.identifier([ 'submissions', 'updatedAt' ])} DESC`);
```
Comment (ktuite) [inline test/unit/data/odata-filter.js:90]: talked about defaulting to ASC

File: lib/data/odata-filter.js (hunk 3 above)
Comment (ktuite) [reply lib/data/odata-filter.js:95]: let's also validate the order

File: test/unit/http/endpoint.js [hunk 6]
```diff
@@ -656,12 +656,6 @@ describe('endpoints', () => {
           .should.be.rejectedWith(Problem, { problemCode: 404.1 });
       });
 
-      it('should reject requests for unsupported OData features', () => {
-        const request = createRequest({ url: '/odata.svc?$orderby=magic' });
-        return odataPreprocessor('json')(null, new Context(request), request)
-          .should.be.rejectedWith(Problem, { problemCode: 501.1 });
```
Comment (ktuite) [inline test/unit/http/endpoint.js:662]: find a different unsupported feature, e.g. $search?? $inlineCount?

File: test/unit/data/odata-filter.js [hunk 7]
```diff
@@ -73,3 +74,48 @@ describe('OData filter query transformer', () => {
...
+      err.message.should.equal('The given OData filter expression references fields not supported by this server: myfield at 0');
+      return true;
+    });
+  });
+
+  it('should ignore unsupported operators and default to DESC', () => {
+    odataOrderBy('__system/updatedAt UP').should.eql(sql`ORDER BY ${sql.identifier([ 'submissions', 'updatedAt' ])} DESC`);
+  });
+
+  it('should add last sort clause to insure stable sort order', () => {
```
Comment (ktuite) [inline test/unit/data/odata-filter.js:118]: also add test of stable id in reverse direction

File: lib/model/query/entities.js [hunk 8]
```diff
@@ -449,7 +449,9 @@ WHERE
   AND entities."deletedAt" IS NULL
   AND entity_defs.current=true
   AND  ${odataFilter(options.filter, odataToColumnMap)}
-ORDER BY entities."createdAt" DESC, entities.id DESC
+${options.orderby ? sql`
```
Comment (ktuite) [inline lib/model/query/entities.js:452]: Hmmmmm, how does it interact with skiptoken?

Comment (ktuite) [reply lib/model/query/entities.js:452]: maybe throw a problem if both are used together?

**Review by ktuite — COMMENTED** [review]
notes from interactive review with Matt

**Review by matthew-white — APPROVED** [review]

**Pull request merged**